libafl_nyx/src/cmplog.rs

```diff
@@ -0,0 +1,220 @@
+{
+    fn pre_exec(&mut self, _state: &mut S, _input: &I) -> Result<(), Error> {
+        unsafe {
+            CMPLOG_ENABLED = 1;
+        }
+        Ok(())
+    }
+
+    fn post_exec(&mut self, state: &mut S, _input: &I, _exit_kind: &ExitKind) -> Result<(), Error> {
+        unsafe {
+            CMPLOG_ENABLED = 0;
+        }
+        if self.add_meta {
+            let meta = state.metadata_or_insert_with(CmpValuesMetadata::new);
+            let rq_data = parse_redqueen_data(&std::fs::read_to_string(self.path.as_ref())?);
+            for event in rq_data.bps {
+                if let Ok(cmp_value) = event.try_into() {
+                    meta.list.push(cmp_value);
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Named for NyxCmpObserver {
+    fn name(&self) -> &Cow<'static, str> {
+        &self.name
+    }
+}
+
+// Based on https://github.com/nyx-fuzz/spec-fuzzer/blob/main/rust_fuzzer/src/runner.rs
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum RedqueenBpType {
+    Str,
+    Cmp,
+    Sub,
+}
+
+impl RedqueenBpType {
+    fn new(data: &str) -> Result<RedqueenBpType, String> {
+        match data {
+            "STR" => Ok(Self::Str),
+            "CMP" => Ok(Self::Cmp),
+            "SUB" => Ok(Self::Sub),
+            _ => Err("Unknown redqueen type".to_string()),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+struct RedqueenEvent {
+    pub addr: u64,
+    pub bp_type: RedqueenBpType,
+    pub size: usize,
+    pub lhs: Vec<u8>,
+    pub rhs: Vec<u8>,
+    pub imm: bool,
+}
+
+impl RedqueenEvent {
+    fn new(line: &str) -> Result<Self, String> {
+        lazy_static! {
+            static ref RE: regex::Regex = regex::Regex::new(
+                r"([0-9a-fA-F]+)\s+(CMP|SUB|STR)\s+(\d+)\s+([0-9a-fA-F]+)-([0-9a-fA-F]+)(\sIMM)?"
+            )
+            .expect("Invalid regex pattern");
+        }
+
+        let captures = RE
+            .captures(line)
+            .ok_or_else(|| format!("Failed to parse Redqueen line: '{line}'"))?;
+
+        let addr_s = captures.get(1).ok_or("Missing address field")?.as_str();
+        let type_s = captures.get(2).ok_or("Missing type field")?.as_str();
+        let size_s = captures.get(3).ok_or("Missing size field")?.as_str();
+        let lhs_s = captures.get(4).ok_or("Missing LHS field")?.as_str();
+        let rhs_s = captures.get(5).ok_or("Missing RHS field")?.as_str();
+        let imm = captures.get(6).is_some_and(|_x| true);
+
+        let addr =
+            u64::from_str_radix(addr_s, 16).map_err(|_| format!("Invalid address: '{addr_s}'"))?;
+        let bp_type = RedqueenBpType::new(type_s)
+            .map_err(|e| format!("Invalid redqueen type: '{type_s}' - {e}"))?;
+        let size = size_s
+            .parse::<usize>()
+            .map_err(|_| format!("Invalid size: '{size_s}'"))?;
+        let lhs =
+            hex::decode(lhs_s).map_err(|e| format!("Failed to decode LHS: '{lhs_s}' - {e}"))?;
```

Comment:
can you not use [from_str_radix](https://doc.rust-lang.org/std/primitive.i64.html#method.from_str_radix)?
this is a new dependency and it's better if we don't add more things else clippy will take more time to run

Reply:
For `RedqueenBpType::Sub` lhs and rhs can be byte arrays of size 32 each. 
In this case `from_str_radix` would not work. I added a helper function to decode the hexstring using an iterator and `u8::from_str_radix`.

Reply:
thank you! good job